File: Learning/src/Questions/BinaryTree/Height.java
package Questions.BinaryTree;

public class Height {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }

    }

    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return 1 + Math.max(maxDepth(root.right), maxDepth(root.left));
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(3);
        TreeNode left = new TreeNode(9);
        TreeNode right = new TreeNode(20);
        TreeNode rightLeft = new TreeNode(15);
        TreeNode rightRight = new TreeNode(7);
        root.left = left;
        root.right = right;
        right.left = rightLeft;
        right.right = rightRight;
        Height height = new Height();
        int result = height.maxDepth(root);
        System.out.println("Height of the tree: " + result);
    }
}

File: DSA/Python/Questions/BinaryTree/Height.py
"""
# Question: Maximum Depth of Binary Tree
# Link: https://leetcode.com/problems/maximum-depth-of-binary-tree/

# Time Complexity: O(N)
# Space Complexity: O(H)

# Algorithm:
# 1. DFS traversal
# 2. Calculate height recursively
# 3. Return maximum height
"""


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class Height:
    def max_depth(self, root: TreeNode) -> int:
        if not root:
            return 0

        left_height = self.max_depth(root.left)
        right_height = self.max_depth(root.right)

        return max(left_height, right_height) + 1


def main():
    solution = Height()
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    print(solution.max_depth(root))


if __name__ == "__main__":
    main()
